File: project/src/main/java/com/expedia/ui/javademo/GenericListAdapter.java
package com.expedia.ui.javademo;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.expedia.ui.javademo.interfaces.HolderAdapterBridge;
import com.expedia.ui.javademo.interfaces.RecyclerViewContract;
import com.expedia.ui.javademo.interfaces.VHClickable;

/**
 * Created by nbirla on 15/11/17.
 */

public class GenericListAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> implements VHClickable, HolderAdapterBridge {

    private FeedItemList mItemList;
    private RecyclerViewContract contract;
    private VHClickable clickCallback;

    public GenericListAdapter(FeedItemList itemList, RecyclerViewContract contract){
        mItemList = itemList;
        this.contract = contract;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        ItemVH viewHolder = contract.createHolder(LayoutInflater.from(parent.getContext()), parent, viewType);
        viewHolder.setVHClickCallback(this);
        return viewHolder;
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        if(!(holder instanceof ItemVH)) return;

        ItemVH itemVH = (ItemVH) holder;
        itemVH.setGLAdapterBridge(this);

        FeedItem feedItem = getItem(position);

        itemVH.bindFeedItem(feedItem);
    }

    private FeedItem getItem(int position){
        return mItemList.get(position);
    }

    public void updateItems(FeedItemList list) {
        mItemList = list;
        notifyDataSetChanged();
    }

    @Override
    public int getItemViewType(int position) {
        return contract.getViewType(getItem(position).getViewType());
    }

    @Override
    public int getItemCount() {
        return mItemList.size();
    }

    public void setVHClickCallback(VHClickable clickCallback) {
        this.clickCallback = clickCallback;
    }

    @Override
    public void onViewHolderClicked(ItemVH holder, View view) {
        if(clickCallback != null){
            clickCallback.onViewHolderClicked(holder, view);
        }
    }

    @Override
    public void notifyVHChanged(int position) {
        notifyItemChanged(position);
    }
}

File: project/src/main/java/com/expedia/ui/recyclerview/GenericListAdapter.kt
package com.expedia.ui.recyclerview

import android.support.v7.widget.RecyclerView
import android.view.ViewGroup
import com.expedia.ui.recyclerview.interfaces.VHClickable
import android.view.LayoutInflater
import android.view.View
import com.expedia.ui.recyclerview.interfaces.HolderAdapterBridge
import com.expedia.ui.recyclerview.interfaces.RecyclerViewContract
import com.expedia.ui.recyclerview.viewholders.VHWrapper
import rx.subjects.PublishSubject

/**
 * Created by nbirla on 15/11/17.
 */
class GenericListAdapter(var itemList: FeedItemList, val contract: RecyclerViewContract) : RecyclerView.Adapter<RecyclerView.ViewHolder>(), HolderAdapterBridge {

    var clickSubject: PublishSubject<VHWrapper>? = null

    override fun onCreateViewHolder(parent: ViewGroup?, viewType: Int): RecyclerView.ViewHolder {

        val viewHolder = contract.createHolder(LayoutInflater.from(parent!!.getContext()), parent, viewType)
        viewHolder.setVHClickCallback(clickSubject)
        return viewHolder

    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder?, position: Int) {

        if (holder !is ItemVH<*>) return

        holder.setGLAdapterBridge(this)

        val feedItem = getItem(position)

        holder.bindFeedItem(feedItem)

    }

    private fun getItem(position: Int): FeedItem<*> {
        return itemList!!.get(position)
    }

    fun updateItems(list: FeedItemList) {
        itemList = list
        notifyDataSetChanged()
    }

    override fun getItemViewType(position: Int): Int {
        return contract.getViewType(getItem(position).getFeedItemViewType());
    }

    override fun getItemCount(): Int {
        return itemList.size
    }

    fun setVHClickSubject(clickSubject: PublishSubject<VHWrapper>){
        this.clickSubject = clickSubject
    }

    override fun notifyVHChanged(position: Int) {
        notifyItemChanged(position);
    }
}